### PR TITLE
Added db client package and related changes

### PR DESCRIPTION
**Pull request by roonyh**

### CODE REVIEW

File: transport/transport.go [hunk 1]
```diff
@@ -3,71 +3,87 @@ package transport
...
-	t.conn.Write(t.d[:8])
+	sz := uint32(len(batch))
 
-	a := uint32(len(batch))
-	hybrid.EncodeUint32(t.d[:4], &a)
-	t.conn.Write(t.d[:4])
+	hybrid.EncodeUint64(t.buf[:8], &id)
+	hybrid.EncodeUint8(t.buf[8:9], &msgType)
+	hybrid.EncodeUint32(t.buf[9:13], &sz)
+	t.conn.Write(t.buf[:13])
```
Comment (thani-sh) [inline transport/transport.go]: Always handle the error. There're a few more places in this file.

File: transport/transport.go [hunk 2]
```diff
@@ -3,71 +3,87 @@ package transport
...
+		readLock:  new(sync.Mutex),
+		buf:       make([]byte, 13),
 	}
 }
 
 // SendBatch writes data to the connection
-func (t *Transport) SendBatch(batch [][]byte, id int64) {
-	t.mtx.Lock()
-	defer t.mtx.Unlock()
+func (t *Transport) SendBatch(batch [][]byte, id uint64, msgType uint8) {
```
Comment (thani-sh) [inline transport/transport.go]: Return an error

File: test/tracker/tracker.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+func main() {
+	var addr string
+	flag.StringVar(&addr, "addr", "localhost:8000", "Host and port of the server <host>:<port>")
+
+	flag.Parse()
+
+	c := client.New()
+	err := c.Connect(addr)
+
```
Comment (thani-sh) [inline test/tracker/tracker.go]: Try not to leave new lines between `err :=` lines and `if err` lines.

File: test/fetcher/fetcher.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	for i := 0; i < 100; i++ {
+		testReqs = append(testReqs, &server.ReqFetch{
+			Database: "kadiyadb",
+			From:     0,
+			To:       uint64(60000000000 * 30),
+			Fields:   []string{"foo", "bar"},
+		})
+	}
+
+	if err != nil {
```
Comment (thani-sh) [inline test/fetcher/fetcher.go]: Test code??

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		if !ok {
+			fmt.Println("Unknown response id")
+			continue
+		}
+
+		ch <- data
+	}
+}
+
+func (c *Client) call(b [][]byte, msgType uint8) [][]byte {
```
Comment (thani-sh) [inline client/client.go]: return an error

File: client/client.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	ch := make(chan [][]byte, 1)
+	id := c.getNextID()
+	c.inflight[id] = ch
+
+	c.tran.SendBatch(b, id, msgType)
```
Comment (thani-sh) [inline client/client.go]: Handle the error

File: client/client.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+	return <-ch
+}
+
+func (c *Client) retrieve(data [][]byte, msgType uint8) ([]*server.Response, error) {
+
+	resData := c.call(data, msgType)
```
Comment (thani-sh) [inline client/client.go]: Handle the error

File: client/client.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+func (c *Client) getNextID() (id uint64) {
+	return atomic.AddUint64(&c.nextID, 1)
+}
+
+// Track tracks kadiyadb points
+func (c *Client) Track(tracks []*server.ReqTrack) ([]*server.Response, error) {
+	data := make([][]byte, len(tracks))
+
+	for i, track := range tracks {
+		data[i], _ = track.Marshal()
```
Comment (thani-sh) [inline client/client.go]: Error

File: client/client.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+	return c.retrieve(data, server.MsgTypeTrack)
+}
+
+// Fetch fetches kadiyadb point data
+func (c *Client) Fetch(fetches []*server.ReqFetch) ([]*server.Response, error) {
+	data := make([][]byte, len(fetches))
+
+	for i, fetch := range fetches {
+		data[i], _ = fetch.Marshal()
```
Comment (thani-sh) [inline client/client.go]: Handle the error

File: test/fetcher/fetcher.go (hunk 4 above)
Comment (roonyh) [inline test/fetcher/fetcher.go]: More like unburied dead code :) Removed. Sorry about that.

**Comment by roonyh** [comment]
Fixed all unhandled errors

File: server/server_test.go [hunk 10]
```diff
@@ -162,61 +127,34 @@ func TestBatch(t *testing.T) {
...
-	}
+	data := make([][]byte, len(tracks))
+	b.ResetTimer()
 
-	data, err := testReq.Marshal()
-	if err != nil {
-		b.Fatal("Marshal gives error", err)
+	// Create a batch
+	for i, req := range tracks {
+		data[i], _ = req.Marshal()
```
Comment (thani-sh) [inline server/server_test.go:158]: One more

**Pull request merged**